### ISSUE TITLE
[5.0] false positive ParadoxicalCondition with match() and enum

### ISSUE THREAD

**Issue by jacekkarczmarczyk** [vimeo/psalm#8812]
https://psalm.dev/r/74b7b0ba5a

Code:
```php
enum SomeType: string
{
    case FOO = 'FOO';
    case BAR = 'BAR';
}

function getSomething(string $moduleString): int
{
    return match ($moduleString) {
        SomeType::FOO->value => 1,
        SomeType::BAR->value => 2,
    };
}
```

Output:
```
ERROR: [ParadoxicalCondition](https://psalm.dev/089) - 13:9 - Condition ($moduleString is string(FOO)) contradicts a previously-established condition ($moduleString is not string(FOO))
```

This did not happen with Psalm 4.x

**Comment by weirdan** [comment]
To solve it we need to correctly infer `SomeType::FOO->value` (currently it's inferred as `'FOO'|'BAR'`: https://psalm.dev/r/71807c8328).
